### PR TITLE
complex ifft implementation

### PR DESCRIPTION
**Pull request by ssmiler**

### CODE REVIEW

File: spqlios/commons.h [hunk 1]
```diff
@@ -5,12 +5,14 @@
...
 #include <cstdio>
 #include <cstdlib>
+#include <cmath>
 #define EXPORT extern "C"
 #define EXPORT_DECL extern "C"
 #else
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <math.h>
```
Comment (ngama75) [inline spqlios/commons.h]: actually, we should move stdio.h, stdlib.h and math.h in and nullptr commons_private.h
This commons.h header will be recursively included by anyone who uses the library, so we should keep it minimal.

File: spqlios/commons_private.h [hunk 2]
```diff
@@ -52,4 +42,11 @@ EXPORT void* spqlios_keep_or_free(void* ptr, void* ptr2);
 #define CPU_SUPPORTS(xxxx) 0
 #endif
 
+uint32_t revbits(uint32_t nbits, uint32_t value);
```
Comment (ngama75) [inline spqlios/commons_private.h:57]:
```suggestion
/** @brief returns the n bits of value in reversed order */
uint32_t revbits(uint32_t nbits, uint32_t value);
```

File: spqlios/commons_private.h [hunk 3]
```diff
@@ -52,4 +42,11 @@ EXPORT void* spqlios_keep_or_free(void* ptr, void* ptr2);
...
+
+/**
+ * @brief this computes the sequence: 0,1/2,1/4,3/4,1/8,5/8,3/8,7/8,...
+ * essentially: the bits of (i+1) in lsb order on the basis (1/2^k) mod 1*/
+double fracrevbits(uint32_t i);
+
 #endif  // SPQLIOS_COMMONS_PRIVATE_H
```
Comment (ngama75) [inline spqlios/commons_private.h]: Let's configure the editor for the new line in the end!

File: spqlios/cplx/cplx_fft_private.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#ifndef SPQLIOS_CPLX_FFT_PRIVATE_H
+#define SPQLIOS_CPLX_FFT_PRIVATE_H
+
+#include "cplx_fft_public.h"
+
+typedef void (*IFFT_FUNCTION)(const CPLX_IFFT_PRECOMP*, void*);
+
+typedef struct cplx_ifft_precomp {
```
Comment (ngama75) [inline spqlios/cplx/cplx_fft_private.h]: we need only the struct, the typedef is already in the public.h

**Review by ngama75 — APPROVED** [review]
Awesome!
Just a few changes. CI should activate when you either merge with / or rebase on main.

File: spqlios/cplx/cplx_ifft_ref.c [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    }
+  }
+  return reps;
+}
+
+EXPORT void* cplx_ifft_precomp_get_buffer(const CPLX_IFFT_PRECOMP* itables, uint32_t buffer_index) {
+  return ((CPLX*)itables->aligned_buffers) + buffer_index * itables->m;
+}
+
+EXPORT void cplx_ifft(const CPLX_IFFT_PRECOMP* itables, void* data) {
```
Comment (ngama75) [inline spqlios/cplx/cplx_ifft_ref.c:282]: on x86, we should use the assembly executor for this function, and fall back on this one on arm64.

Comment (ssmiler) [reply spqlios/cplx/cplx_ifft_ref.c:282]: Ok

**Review by MGeorgie — APPROVED** [review]
good to merge

**Pull request merged**